File: src/main/kotlin/Lesson_4/MDIBT_problems.java
package Lesson_4;


import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MDIBT_problems {
    public static List<Integer> findMode(List<Integer> array) {
        Map<Integer, Integer> countMap = new HashMap<>();
        for (int num : array) {
            countMap.put(num, countMap.getOrDefault(num, 0) + 1);
        }

        int maxCount = 0;
        List<Integer> modes = new ArrayList<>();

        for (Map.Entry<Integer, Integer> entry : countMap.entrySet()) {
            int count = entry.getValue();
            if (count > maxCount) {
                maxCount = count;
                modes.clear();
                modes.add(entry.getKey());
            } else if (count == maxCount) {
                modes.add(entry.getKey());
            }
        }

        return modes;
    }

    public static List<Integer> sortArray(List<Integer> array) {
        List<Integer> sortedArray = new ArrayList<>(array);
        Collections.sort(sortedArray);
        return sortedArray;
    }

    public static double findMedian(List<Integer> array) {
        List<Integer> sortedArray = new ArrayList<>(array);
        Collections.sort(sortedArray);
        int count = sortedArray.size();

        if (count % 2 == 1) {
            return (double) sortedArray.get(count / 2);
        } else {
            int middle1 = sortedArray.get(count / 2 - 1);
            int middle2 = sortedArray.get(count / 2);
            return (middle1 + middle2) / 2.0;
        }
    }

    public static double findDispersiya(List<Integer> array) {
        int n = array.size();
        double mean = array.stream().mapToInt(Integer::intValue).average().orElse(0.0);
        double sum = 0.0;

        for (int value : array) {
            sum += Math.pow(value - mean, 2);
        }

        return sum / n;
    }

    public static double findOrtacha(List<Integer> array) {
        return array.stream().mapToInt(Integer::intValue).average().orElse(0.0);
    }

    public static List<Double> findRang(List<Integer> array) {
        List<Integer> sortedArray = new ArrayList<>(array);
        Collections.sort(sortedArray);
        List<Double> rang = new ArrayList<>();

        for (int i = 0; i < sortedArray.size(); ) {
            int j = i;
            while (j < sortedArray.size() && sortedArray.get(i).equals(sortedArray.get(j))) {
                j++;
            }

            double rangValue = (i + j + 1) / 2.0;
            for (int k = i; k < j; k++) {
                rang.add(rangValue);
            }

            i = j;
        }

        return rang;
    }

    public static void main(String[] args) {
        List<Integer> array = new ArrayList<>();
        Collections.addAll(array, 3, 0, 5, 1, 1, 5, 3);
        List<Double> rangValues = findRang(array);
        List<Integer> modes = findMode(array);

        System.out.print("Qator elementlari: ");
        for (int item : array) {
            System.out.print(item + " ");
        }
        System.out.println();
        System.out.print("Variatsion qator : ");
        List<Integer> sortedArray = sortArray(array);
        for (int item : sortedArray) {
            System.out.print(item + " ");
        }
        System.out.println();
        System.out.print("Mode: ");
        for (int mode : modes) {
            System.out.print(mode + " ");
        }
        System.out.println();
        System.out.println("Mediana: " + findMedian(array));
        System.out.println("Dispersiya: " + findDispersiya(array));
        System.out.println("Razmax: " + (Collections.max(array) - Collections.min(array)));
        System.out.println("O'rtacha qiymat: " + findOrtacha(array));
        System.out.print("Rang values: ");
        for (double item : rangValues) {
            System.out.print(item + " ");
        }
    }
}

File: src/main/kotlin/Lesson_4/MDIBT_PROBLEMS.kt
package Lesson_4

fun findMode(array: List<Int>): List<Int> {
    val countMap = mutableMapOf<Int, Int>()

    for (num in array) {
        countMap[num] = countMap.getOrDefault(num, 0) + 1
    }

    var maxCount = 0
    val modes = mutableListOf<Int>()

    for ((key, count) in countMap) {
        if (count > maxCount) {
            maxCount = count
            modes.clear()
            modes.add(key)
        } else if (count == maxCount) {
            modes.add(key)
        }
    }

    return modes
}

fun sortArray(array: List<Int>): List<Int> {
    return array.sorted()
}

fun findMedian(array: List<Int>): Double {
    val sortedArray = array.sorted()
    val count = sortedArray.size

    return if (count % 2 == 1) {
        sortedArray[count / 2].toDouble()
    } else {
        val middle1 = sortedArray[count / 2 - 1]
        val middle2 = sortedArray[count / 2]
        (middle1 + middle2) / 2.0
    }
}

fun findDispersiya(array: List<Int>): Double {
    val n = array.size
    val mean = array.average()
    var sum = 0.0

    for (value in array) {
        sum += Math.pow(value - mean, 2.0)
    }

    return sum / n
}

fun findOrtacha(array: List<Int>): Double {
    return array.average()
}

fun findRang(array: List<Int>): List<Double> {
    val sortedArray = array.sorted()
    val rang = mutableListOf<Double>()

    var i = 0
    while (i < sortedArray.size) {
        var j = i
        while (j < sortedArray.size && sortedArray[i] == sortedArray[j]) {
            j++
        }

        val rangValue = (i + j + 1) / 2.0
        repeat(j - i) {
            rang.add(rangValue)
        }

        i = j
    }

    return rang
}

fun main() {
    val array = mutableListOf(3, 0, 5, 1, 1, 5, 3)
    val rangValues = findRang(array)
    val modes = findMode(array)

    print("Qator elementlari: ")
    array.forEach { print("$it ") }
    println()
    print("Variatsion qator : ")
    val sortedArray = sortArray(array)
    sortedArray.forEach { print("$it ") }
    println()
    print("Mode: ")
    modes.forEach { print("$it ") }
    println()
    println("Mediana: ${findMedian(array)}")
    println("Dispersiya: ${findDispersiya(array)}")
    println("Razmax: ${array.maxOrNull()!! - array.minOrNull()!!}")
    println("O'rtacha qiymat: ${findOrtacha(array)}")
    print("Rang values: ")
    rangValues.forEach { print("$it ") }
}
